### ISSUE TITLE
chi-squred test ?

### ISSUE THREAD

**Issue by behrica** [generateme/fastmath#22]
I would like to select relevant TFIDF features for machine learining.
One way to do this is:

https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.chi2.html

I could not really see if `fastmath` supports this currently ?

**Comment by behrica** [comment]
same question for:

https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html#sklearn.feature_selection.f_classif

**Comment by genmeblog** [comment]
chi2 test is implemented but private. I'll push a snapshot today.

**Comment by behrica** [comment]
yes, I saw the method in teh code,
There are slighly differences in te data types.

The sklearn  chi2 works with 2D double (X)  and 1D (int) y.

That is what I would need.

**Comment by behrica** [comment]
yes, I saw the method in teh code,
There are slighly differences in te data types.

The sklearn  chi2 works with 2D double (X)  and 1D (int) y.

That is what I would need.

**Comment by genmeblog** [comment]
I can expose only this one (I mean 1d,1d). I don't know how to make it work for 2d,1d case easily (I mean, I don't get this python code).

`f_classif` is based on one-way anova which require much more stuff not available currently.

**Comment by genmeblog** [comment]
1d case deployed.

**Comment by genmeblog** [comment]
@behrica I introduced `chisq-test` as it works in Julia and R. Here is the guidance how to use it for your case.

Suppose your data are (5 observation, 2 variables, 2 categories):

```
(def X [[2 3] [9 1] [3 4] [5 5] [9 9]])
(def y [0 1 1 0 0])
```

```python
>>> X = [[2, 3], [9, 1], [3, 4], [5, 5], [9, 9]]
>>> y = [0, 1, 1, 0, 0]
>>> s.chi2(X,y)
(array([0.0952381 , 2.73484848]), array([0.75762072, 0.09818081]))
```

Python applies chi2 test for each variable separately. First of all we need to convert categories into probabilities.

```clojure
(def probs (-> (frequencies y)
             (update-vals #(double (/ % (count y))))))

probs ;; => {0 0.6, 1 0.4}
```

Now we need to group and sum each variable by categories, here is the first variable:

```clojure
(def variable-1 (-> (->> (map first X) ;; first variable
                       (map vector y) ;; join with categories
                       (group-by first)) ;; group by categories
                  (update-vals #(sum (map second %))))) ;; sum by categories
;; => {0 16.0, 1 12.0}
```

This transformation gives a input for the goodness of fit chi2 test:

```clojure
(-> (chisq-test (vals variable-1) {:p (vals probs)})
    (select-keys [:chi2 :p-value]))
;; => {:chi2 0.09523809523809401, :p-value 0.7576207236833981}
```

The same for second variable:

```clojure
(def variable-2 (-> (->> (map second X) ;; second variable
                       (map vector y) ;; join with categories
                       (group-by first)) ;; group by categories
                  (update-vals #(sum (map second %))))) ;; sum by categories

variable-2
;; => {0 17.0, 1 5.0}

(-> (chisq-test (vals variable-2) {:p (vals probs)})
    (select-keys [:chi2 :p-value]))
;; => {:chi2 2.7348484848484844, :p-value 0.0981808126896524}
```

I need to find how to make it for One Way Anova (implemented in fastmath already).

**Comment by genmeblog** [comment]
One Way Anova works without any changes:

```python
>>> s.f_oneway([1,2,3,4,5],[5,6,5,6,5],[7,6,7,6,7])
(array([16.25806452]), array([0.00038369]))
```

Full data from one-way Anova (`:F` and `:p-value` - are the same as in Python)

```clojure
(one-way-anova-test [[1,2,3,4,5],[5,6,5,6,5],[7,6,7,6,7]])
;; => {:DFe 12,
;;     :stat 16.258064516129032,
;;     :F 16.258064516129032,
;;     :DFt 2,
;;     :p-value 3.8369233276935866E-4,
;;     :df [2 12],
;;     :MSt 16.799999999999997,
;;     :SSe 12.399999999999999,
;;     :MSe 1.0333333333333332,
;;     :SSt 33.599999999999994,
;;     :N (5 5 5)}
```

**Comment by behrica** [comment]
Thanks   , will try it soon

**Comment by genmeblog** [comment]
Solved in 2.2.0
